### PR TITLE
[CALCITE-3961] VolcanoPlanner.prunedNodes information is lost when duplicate relNode is discarded

### PR DESCRIPTION
**Pull request by hbtoo**
https://issues.apache.org/jira/browse/CALCITE-3961

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java [hunk 1]
```diff
@@ -361,26 +361,16 @@ void mergeWith(
 
     Set<RelNode> parentRels = new HashSet<>(parents);
     for (RelNode otherRel : otherSet.rels) {
-      if (planner.prunedNodes.contains(otherRel)) {
-        continue;
-      }
-
-      boolean pruned = false;
       if (parentRels.contains(otherRel)) {
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java:364]: Can you also add a check `otherRel` is not `spool` and is not an enforcer before `parentRels.contains(otherRel)`? I think we can't prune them in that case, what do you think?

Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java:364]: Maybe we can also remove https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java#L127-L130, they look duplicate with this part.

Comment (hbtoo) [reply core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java]: done

File: core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java [hunk 2]
```diff
@@ -940,11 +941,24 @@ void reregister(
...
+    if (!prunedNodes.contains(rel)) {
+      addRelToSet(rel, set);
+    }
+  }
+
+  /**
+   * Update relNode related information when two identical relNodes are
+   * de-duped.
+   */
+  private void mergeIdenticalRels(RelNode to, RelNode from) {
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java:958]: from, to is confusing. Can we use equivRel, rel instead? And the comment doesn't elaborate what does it do.

Comment (hbtoo) [reply core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java]: I changed it to RelNode rel, RelNode duplicate, where duplicate is the one that's about to be discarded.

**Pull request closed without merging**